docs: improve service-repository architecture description in readme (#268)
RonasIT/laravel-helpers: documentation/service-repository.md

[<< Traits][1]
[Services >>][2]

# Service-Repository Architecture

The package provides base classes for implementing the service-repository pattern. This pattern separates business logic (services) from data access (repositories), making the codebase more maintainable, testable, and easier to scale.

This results in a clean layered architecture:

```
Controller → Service → Repository → Model
```

- **Service** handles business logic, orchestrates application workflows, and delegates data operations to the repository.

> - **Repository** encapsulates all database queries and CRUD operations.

Reviewer: ```suggestion
- **Repository** encapsulates work with data storage and provide base CRUD operations.
```

## Setting Up

> ### 1. Create a Repository

Reviewer: let's add a small sentence about entity generator before the manual creation

Extend `BaseRepository` and set the model in the constructor:

```php
use RonasIT\Support\Repositories\BaseRepository;

final class UserRepository extends BaseRepository
{
    public function __construct()
    {
        $this->setModel(User::class);
    }
}
```

`BaseRepository` uses `EntityControlTrait`, which provides all CRUD and search methods automatically.

### 2. Create a Service

Extend `EntityService` and associate it with a repository:

```php
use RonasIT\Support\Services\EntityService;

final class UserService extends EntityService
{
    public function __construct()
    {
        $this->setRepository(UserRepository::class);
    }
}
```

> `EntityService` uses `__call()` to delegate method calls to the repository. If a repository method returns `$this` (for chaining), the service returns itself instead, allowing seamless method chaining through the service layer.

Reviewer: ```suggestion
`EntityService` uses `__call()` to delegate method calls to the repository in case it not exists in the service class.

If a repository method returns `$this` (for chaining), the service returns itself instead, allowing seamless method chaining through the service layer.
```

> 

Reviewer: Let's describe method redefinition and add example for it

---

## CRUD Operations

All methods below are available on both the repository and the service (via delegation).

> The `$where` parameter accepts a primary key value (`int` or `string`) or an associative array of conditions:

Reviewer: ```suggestion
The most methods support the mixed input for the condition. It accepts a primary key value or an associative array of conditions:
```

```php
$this->find(1);
$this->first(['email' => 'user@example.com']);
$this->update(1, ['name' => 'New Name']);

> $this->update(['email' => 'user@example.com'], ['name' => 'New Name']);

Reviewer: ```suggestion
$this->first($id) === $this->first(['id' => $id]);
$this->first(['email' => $email, 'is_active' => true]);

$this->update($id, ['name' => 'New Name']);
$this->update(['email' => 'user@example.com'], ['name' => 'New Name']);
```

```

### Create

| Method | Description |
|--------|-------------|

> | `create(array $data): Model` | Create a new entity and return it with loaded relations |

Reviewer: ```suggestion
| `create(array $data): Model` | Create and return a new entity |
```

> | `insert(array $data): bool` | Mass insert entities with automatic timestamps |

Reviewer: ```suggestion
| `insert(array $data): bool` | Mass insert entities, the timestamps fields filling accordingly to the related model |
```

| `insertOrIgnore(array $data): int` | Mass insert rows, silently skipping duplicate key errors. Returns count of inserted rows |
| `firstOrCreate(array\|int\|string $where, array $data = []): Model` | Get the first entity matching the condition or create a new one |
| `updateOrCreate(array\|int\|string $where, array $data): Model` | Update an existing entity or create a new one |

### Read

| Method | Description |
|--------|-------------|
| `find(int\|string $id): ?Model` | Find an entity by primary key |
| `findBy(string $field, mixed $value): ?Model` | Find an entity by a specific field value |

> | `first(array\|int\|string $where = []): ?Model` | Get the first entity matching the condition |

Reviewer: ```suggestion
| `first(array\|int\|string $where = []): ?Model` | Find the first entity matching the condition |
```

> | `last(array\|int\|string $where = [], string $column = 'created_at'): ?Model` | Get the last entity matching the condition, ordered by `$column` |

Reviewer: ```suggestion
| `last(array\|int\|string $where = [], string $column = 'created_at'): ?Model` | Find the last entity matching the condition, ordered by `$column` |
```

> | `get(array\|int\|string $where = []): Collection` | Get all entities matching the condition |

Reviewer: ```suggestion
| `get(array\|int\|string $where = []): Collection` | Get a list of all entities matching the condition |
```

> | `getByList(array $values, ?string $field = null): Collection` | Get entities whose `$field` value is in `$values` (defaults to primary key) |

Reviewer: ```suggestion
| `getByList(array $values, ?string $field = null): Collection` | Get the list of entities whose `$field` (defaults to primary key) value is in the list of `$values` |
```

> | `all(): Collection` | Get all entities without conditions |

Reviewer: ```suggestion
| `all(): Collection` | Get the list of all entities without conditions |
```

| `exists(array\|int\|string $where): bool` | Check entity existence by condition or primary key |
| `existsBy(string $field, mixed $value): bool` | Check entity existence by a specific field value |
| `count(array\|int\|string $where = []): int` | Count entities by condition or primary key |

> | `countByList(array $values, ?string $field = null): int` | Count entities whose `$field` value is in `$values` (defaults to primary key) |

Reviewer: ```suggestion
| `countByList(array $values, ?string $field = null): int` | Count entities whose `$field` (defaults to primary key) value is in `$values` list |
```

| `chunk(int $limit, Closure $callback, array $where = []): void` | Process entities in chunks ordered by primary key |

### Update

| Method | Description |
|--------|-------------|

> | `update(array\|int\|string $where, array $data): ?Model` | Update a single entity and return it |

Reviewer: ```suggestion
| `update(array\|int\|string $where, array $data): ?Model` | Update and return a single entity |
```

| `updateMany(array\|int\|string $where, array $data): int` | Update all entities matching the condition. Returns count of updated rows |

> | `updateByList(array $values, array $data, ?string $field = null): int` | Update entities whose `$field` value is in `$values` (defaults to primary key). Returns count of updated rows |

Reviewer: ```suggestion
| `updateByList(array $values, array $data, ?string $field = null): int` | Update entities whose `$field` (defaults to primary key) value is in the `$values` list. Returns count of updated rows |
```

### Delete

| Method | Description |
|--------|-------------|
| `delete(array\|int\|string $where): int` | Delete entities by condition or primary key. Returns count of deleted rows |

> | `deleteByList(array $values, ?string $field = null): int` | Delete entities whose `$field` value is in `$values` (defaults to primary key). Returns count of deleted rows |

Reviewer: ```suggestion
| `deleteByList(array $values, ?string $field = null): int` | Delete entities whose `$field` (defaults to primary key) value is in the `$values` list. Returns count of deleted rows |
```

| `truncate(): self` | Remove all rows from the table |

### `force(bool $value = true): self`

> By default, `create()` and `update()` only fill attributes listed in the model's `$fillable` array. Use `force()` to bypass fillable restrictions and write all model fields (including guarded ones):

Reviewer: ```suggestion
The force mode makes different affect for the different cases:
1. For the `delete` operation, force mode makes force delete operation even the model has SoftDeletes trait
2. For the create/update operations force mode allows to fill the guarded or non defined fields of the model.
```

```php
$this->force()->create($data);
$this->force()->update($where, $data);

> $this->force()->updateMany($where, $data);

Reviewer: ```suggestion
$this->force()->delete($where);
```

$this->force()->updateByList($ids, $data);
```

> [!NOTE]
> `force()` is chainable and resets automatically after the next query.

---

## Soft Delete Support

> For models using Laravel's `SoftDeletes` trait, the following methods are available:

Reviewer: let's also add information about the package supports work with the soft deleting and non force delete records in this case

### Scoping

| Method | Description |
|--------|-------------|

> | `withTrashed(bool $enable = true): self` | Include soft-deleted entities in query results |

Reviewer: ```suggestion
| `withTrashed(bool $enable = true): self` | Include soft-deleted entities into operations |
```

> | `onlyTrashed(bool $enable = true): self` | Return only soft-deleted entities |

Reviewer: ```suggestion
| `onlyTrashed(bool $enable = true): self` | Exclude non soft-deleted entities from operations |
```

> [!NOTE]
> Both methods are chainable and apply to the next query only.

### Restore

| Method | Description |
|--------|-------------|
| `restore(array\|int\|string $where): int` | Restore soft-deleted entities by condition or primary key. Returns count of restored rows |

> | `restoreByList(array $values, ?string $field = null): int` | Restore soft-deleted entities whose `$field` value is in `$values` (defaults to primary key). Returns count of restored rows |

Reviewer: ```suggestion
| `restoreByList(array $values, ?string $field = null): int` | Restore soft-deleted entities whose `$field` (defaults to primary key) value is in the `$values` list. Returns count of restored rows |
```

### Examples

```php
// Include soft-deleted entities in results
$this->withTrashed()->get();

// Get only soft-deleted entities
$this->onlyTrashed()->get();

// Restore by condition
$this->restore(['status' => 'banned']);

// Restore by list of IDs
$this->restoreByList([1, 2, 3]);

// Permanently delete (bypass soft delete)
$this->force()->delete($where);
$this->force()->deleteByList([1, 2, 3]);
```

---

## Eager Loading

> 

Reviewer: let's add link to the eager loading documentation and describe methods, supported eager loading

| Method | Description |
|--------|-------------|
| `with(array\|string $relations): self` | Sets relations to eager load on the next query. Resets after the query |

> | `withCount(array\|string $relations): self` | Loads relation counts. Supports dot notation. Resets after the query |

Reviewer: let's add note about resetting as you made in other cases

```suggestion
| `with(array\|string $relations): self` | Sets relations to eager load on the next query. |
| `withCount(array\|string $relations): self` | Loads relation counts. Supports dot notation. |
```

---

## Search and Filtering

`SearchTrait` (included via `EntityControlTrait`) provides a search pipeline with automatic filter resolution, manual filter methods, ordering, and pagination.

### Basic Usage

> 

Reviewer: Base search implementation consists of 2 code methods:
1. `searchQuery` which save the filters list, prepare query object and apply predefined filters
2. `getSearchResults` generating paginated result of the prepared query

```php
public function search(array $filters): LengthAwarePaginator
{
    return $this
        ->searchQuery($filters)

>         ->filterByQuery(['name', 'email'])

Reviewer: ```suggestion
```

        ->getSearchResults();
}
```

> ### `searchQuery(array $filters): self`

Reviewer: ```suggestion
### Predefined filters
```

> Initializes the query with eager loading and soft-delete scoping from `$filters`, then auto-applies all non-reserved filters by suffix:

Reviewer: The `searchQuery` method automatically apply the predefined filters to the query object. Each key of the `filters` argument will be interpreted as a filter by full matching (in case it hasn't reserved postfix), e.g.

//Add example about ->search(['name' => 'Foo']) //will return all records which has `foo` value in the `name` field

> | Suffix | Operator | Example filter key | Example filter value |

Reviewer: ### Reserved postfixes

In case some of the `$filters` keys contains the special postfix - it will be automatically interpreted as a predefined filter.

|--------|----------|--------------------|----------------------|
| `_in_list` | `whereIn` | `status_in_list` | `['active', 'pending']` |
| `_not_in_list` | `whereNotIn` | `status_not_in_list` | `[1, 2]` |
| `_gte` | `>=` | `age_gte` | `18` |
| `_gt` | `>` | `price_gt` | `100` |
| `_lte` | `<=` | `age_lte` | `65` |
| `_lt` | `<` | `price_lt` | `1000` |
| `_from` | `>=` | `created_at_from` | `'2024-01-01'` |

> | `_to` | `<=` | `created_at_to` | `'2024-12-31'` |

Reviewer: Let's simplify the table and add few examples after it

```suggestion
| Postfix | Operator | Description |
|--------|----------|--------------------|----------------------|
| `_in_list` | `whereIn` | Field value should be one of provided | 
```

| *(none)* | `=` | `status` | `'active'` |

### Reserved Filter Names

> These filter keys are handled internally and should not be used as field filters: `with`, `with_count`, `with_trashed`, `only_trashed`, `query`, `order_by`, `all`, `per_page`, `page`, `desc`.

Reviewer: ```suggestion
The search uses next reserved fields in the logic: `with`, `with_count`, `with_trashed`, `only_trashed`, `query`, `order_by`, `all`, `per_page`, `page`, `desc`.

We highly recommend do not use these filters to implement a custom logic.
```

To add custom reserved filter names, call `setAdditionalReservedFilters()` in the repository constructor:

```php
$this->setAdditionalReservedFilters('coach_id', 'contact_id');

> ```

Reviewer: let's move this text to this block https://github.com/RonasIT/laravel-helpers/pull/268/changes#r3264430848

> [!NOTE]
> These keys will be skipped by the auto-filter logic and can be handled manually in a custom `search()` method.

### Manual Filter Methods

> Use these after `searchQuery()` for fine-grained control:

Reviewer: ```suggestion
If you need to extend the predefined filters logic - use filters manually with your own logic. Each filter support the fluent syntax.
```

| Method | Description |
|--------|-------------|

> | `filterBy(string $field, ?string $filterName = null): self` | Exact match filter. Supports dot notation for relations (e.g., `role.name`) |

Reviewer: let's remove signature from the table

```suggestion
| `filterBy` | Exact match filter. Supports dot notation for relations (e.g., `role.name`) |
```

| `filterByList(string $field, ?string $filterName = null): self` | `whereIn` filter |
| `filterByQuery(array $fields, string $mask = "'%{{ value }}%'"): self` | `LIKE` search across multiple fields. Supports relation fields via dot notation |
| `filterGreater(string $field, bool $isStrict = true, ?string $filterName = null): self` | `>` / `>=` filter. Reads from `$filters[$filterName]`, defaults to `'from'` |
| `filterLess(string $field, bool $isStrict = true, ?string $filterName = null): self` | `<` / `<=` filter. Reads from `$filters[$filterName]`, defaults to `'to'` |
| `filterValue(string $field, string $sign, mixed $value): self` | Applies a comparison condition with a given value directly. Skips if empty |
| `orderBy(?string $default = null, bool $defaultDesc = false): self` | Applies ordering by `$filters['order_by']`. Supports dot notation for relation fields |

### Pagination

> `getSearchResults()` returns a `LengthAwarePaginator`. Pagination is controlled via filters:

Reviewer: ```suggestion
The pagination is out of the box feature, provided by `getSearchResults()` method. Pagination works using next predefined filters:
```

| Filter | Description | Default |
|--------|-------------|---------|

> | `per_page` | Items per page | `config('defaults.items_per_page')` |

Reviewer: ```suggestion
| `per_page` | Items per page | Config value `defaults.items_per_page` |
```

| `page` | Current page | `1` |

> | `all` | Return all results without pagination | `false` |

Reviewer: ```suggestion
| `all` | Return all results without pagination, wrapped into the pagination structure | `false` |
```

| `order_by` | Field to sort by | Primary key |
| `desc` | Sort descending | `false` |

> ### Advanced Methods

Reviewer: not sure we need to document the "local" only methods
